js/web/lib/wasm/jsep/webgpu/ops/concat.ts

```diff
@@ -153,10 +164,27 @@ const createConcatProgramInfo = (inputs: readonly TensorView[], axis: number): P
 };
 
 export const concat = (context: ComputeContext, attributes: ConcatAttributes): void => {
-  validateInputs(context.inputs);
+  // find a none zero tensor to determine the output shape
+  // Choose input with max rank if  all input tensors are zero size to make the output shape independent of the order of
+  // the inputs.
+  let referenceIndex = context.inputs.findIndex(input => ShapeUtil.size(input.dims) > 0);
+  if (referenceIndex === -1) {
+    referenceIndex =
+        context.inputs.map(input => input.dims.length)
+            .reduce((maxRankIndex, rank, index, array) => rank > array[maxRankIndex] ? index : maxRankIndex, 0);
+  }
+
+  validateInputs(context.inputs, referenceIndex);
+  const inputShape = context.inputs[referenceIndex].dims;
+  const adjustedAxis = attributes.axis + (attributes.axis < 0 ? inputShape.length : 0);
+  const outputShape = computeOutputShape(context.inputs, adjustedAxis, referenceIndex);
   // 0 length tensors are valid for concat, remove them
   const nonEmptyInputs = context.inputs.filter(input => ShapeUtil.size(input.dims) > 0);
```

Comment:
seems there are a few duplicated calculation (correct me if I am wrong) for calculating referenceIndex, the output shape and nonEmptyInputs. Maybe use a helper function to calculate all to reduce the complexity?

Reply:
Removed computeOutputShae function and moved the validation to validateInputs function.